In [1]:
import pandas as pd
import plotly.express as pe

In [2]:
link = "https://raw.githubusercontent.com/techficent/PythonDABatch/main/titanic.csv"
df = pd.read_csv(link)
df

Unnamed: 0  PassengerId  Survived  Pclass  \
0             0            1         0       3   
1             1            2         1       1   
2             2            3         1       3   
3             3            4         1       1   
4             4            5         0       3   
..          ...          ...       ...     ...   
886         886          887         0       2   
887         887          888         1       1   
888         888          889         0       3   
889         889          890         1       1   
890         890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 13 columns]

### Average, total , minimum & maximum from data

(Aggregations)

In [3]:
df.Fare.sum()

28693.9493

In [4]:
df.Fare.min()

0.0

In [5]:
df.Fare.max()

512.3292

In [6]:
df.Fare.mean()

32.204207968574636

### multiple columns

In [8]:
df[         ["Age", "Fare"]            ].mean() 

Age     29.699118
Fare    32.204208
dtype: float64

### multiple columns, multiple aggregations???

### minimum & maximum of both Age and Fare column together!!!

In [9]:
df[        ["Age","Fare"]           ].agg(           ["min","max"]            )

Age      Fare
min   0.42    0.0000
max  80.00  512.3292

In [10]:
df[        ["Age","Fare"]           ].agg(           ["mean","sum"]            )

Age          Fare
mean     29.699118     32.204208
sum   21205.170000  28693.949300

#### locating particular row numbers and column numbers

In [11]:
df.iloc[     [10,12,21]   ,  [3,7]          ]

Pclass  SibSp
10       3      1
12       3      0
21       2      0

loc : index and column name

In [12]:
df.loc[       [10,19,21],  ["Age", "Fare"]      ]

Age    Fare
10   4.0  16.700
19   NaN   7.225
21  34.0  13.000

## sequence of row numbers

In [13]:
df.loc[       20:30        ,   ["Age","Fare"]           ]

Age      Fare
20  35.0   26.0000
21  34.0   13.0000
22  15.0    8.0292
23  28.0   35.5000
24   8.0   21.0750
25  38.0   31.3875
26   NaN    7.2250
27  19.0  263.0000
28   NaN    7.8792
29   NaN    7.8958
30  40.0   27.7208

#### chained query: multiple operations one after the other, left to right in sequence

#minimum and maximum age and fare for data between rows 20 and 30

In [14]:
df.loc[       20:30        ,   ["Age","Fare"]           ].agg(['min','max'])

Age     Fare
min   8.0    7.225
max  40.0  263.000

## Find average, min, max and total Fare for passengers in row 10, 21, 45, 67 and 90


In [15]:
df.loc[      [10,21,45,67,90], ["Fare"]           ].agg(    ['min',"max","sum","mean"]        )

Fare
min    8.05000
max   16.70000
sum   53.95830
mean  10.79166

In [18]:
rows=[10,21,45,67,90]
col=["Fare"]
ops=['min',"max","sum","mean"]

result = df.loc[   rows   ,     col       ].agg(     ops       )

result

Fare
min    8.05000
max   16.70000
sum   53.95830
mean  10.79166

## use bar chart where one attribute is categorical and other is real-value

In [19]:
pe.bar(     x=result.index,    y=result.Fare             )

### grouping of data


Average Fare for all passengers!

df.Fare.mean()


### Pclass1 ,  Pclass 2,  Pclass 3

891:       

        Pclass 1 : mean 
        Pclass 2 : mean
        Pclass 3 : mean

In [ ]:
## combine-apply 

In [20]:
df.groupby(     ['Pclass']     ).Fare.mean() 

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64

### grouping operations has 3 steps

### 1) split the data (use groupby)
### 2) select the columns
### 3) select aggregate operations

In [23]:
grp=['Pclass']

cols = ["Age", "Fare"]

ops = ['mean','sum']


df.groupby(    grp      ) [      cols       ].agg(        ops       )

Age                Fare            
             mean      sum       mean         sum
Pclass                                           
1       38.233441  7111.42  84.154687  18177.4125
2       29.877630  5168.83  20.662183   3801.8417
3       25.140620  8924.92  13.675550   6714.6951